### PR TITLE
fix(runtime): make some things more stable (partial jsc debug build)

### PR DESCRIPTION
**Pull request by paperclover**
This PR has been working to getting bun-debug working with JSC debug assertions. instead of fully shipping that, this simply is just some of the important fixes i've found in this project. The final version of this project will involve a handful of changes to CI, Build, and CI+Publishing for oven-sh/webkit in order to make it so JSC debug is the default for the dev environment.

### CODE REVIEW

File: src/bun.js/bindings/JSBufferList.cpp [hunk 1]
```diff
@@ -132,13 +131,14 @@ JSC::JSValue JSBufferList::_getString(JSC::VM& vm, JSC::JSGlobalObject* lexicalG
     }
     if (n < len) {
         JSString* firstHalf = JSC::jsSubstring(lexicalGlobalObject, str, 0, n);
-        iter->set(vm, this, JSC::jsSubstring(lexicalGlobalObject, str, n, len - n));
+        m_deque.first().set(vm, this, JSC::jsSubstring(lexicalGlobalObject, str, n, len - n));
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/JSBufferList.cpp:134]: Do we need to check that m_dequeue is not empty?

File: src/bun.js/bindings/JSDOMWrapperCache.h [hunk 2]
```diff
@@ -215,6 +215,10 @@ template<typename DOMClass> inline void setSubclassStructureIfNeeded(JSC::JSGlob
     JSC::VM& vm = lexicalGlobalObject->vm();
     auto scope = DECLARE_THROW_SCOPE(vm);
 
+    // If the new target isn't actually callable
+    if (!newTarget->isCallable())
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/JSDOMWrapperCache.h]:
```suggestion
    if (UNLIKELY(!newTarget->isCallable()))
```

File: src/bun.js/bindings/Process.lut.h [hunk 3]
```diff
@@ -304,7 +304,6 @@ static const struct HashTableValue processObjectTableValues[66] = {
    { "hrtime"_s, static_cast<unsigned>(PropertyAttribute::PropertyCallback), NoIntrinsic, { HashTableValue::LazyPropertyType, constructProcessHrtimeObject } },
    { "isBun"_s, static_cast<unsigned>(PropertyAttribute::PropertyCallback), NoIntrinsic, { HashTableValue::LazyPropertyType, constructIsBun } },
    { "kill"_s, static_cast<unsigned>(PropertyAttribute::Function), NoIntrinsic, { HashTableValue::NativeFunctionType, Process_functionKill, 2 } },
-   { "mainModule"_s, ((static_cast<unsigned>(PropertyAttribute::ReadOnly|PropertyAttribute::Builtin|PropertyAttribute::Accessor|PropertyAttribute::Function)) & ~PropertyAttribute::Function) | PropertyAttribute::Builtin, NoIntrinsic, { HashTableValue::BuiltinGeneratorType, processObjectMainModuleCodeGenerator, 0 } },
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/Process.lut.h]: please implement `process.mainModule`

It is a getter for `require.main` I think

File: src/bundler/entry_points.zig [hunk 4]
```diff
@@ -156,6 +157,25 @@ pub const ClientEntryPoint = struct {
     }
 };
 
+const QuoteEscapeFormat = struct {
+    data: []const u8,
+
+    pub fn format(self: QuoteEscapeFormat, comptime _: []const u8, _: std.fmt.FormatOptions, writer: anytype) !void {
+        var i: usize = 0;
+        while (std.mem.indexOfAnyPos(u8, self.data, i, "\"\n\\")) |j| : (i = j + 1) {
```
Comment (Jarred-Sumner) [inline src/bundler/entry_points.zig:165]: nice

File: src/deps/lol-html.zig [hunk 5]
```diff
@@ -582,7 +582,7 @@ pub const HTMLString = extern struct {
 
     pub fn toString(this: HTMLString) bun.String {
         const bytes = this.slice();
-        if (bun.strings.isAllASCII(bytes)) {
+        if (bytes.len > 0 and bun.strings.isAllASCII(bytes)) {
```
Comment (Jarred-Sumner) [inline src/deps/lol-html.zig:585]: is this necessary?

File: src/js/_codegen/helpers.ts [hunk 6]
```diff
@@ -42,3 +42,12 @@ export function resolveSyncOrNull(specifier: string, from: string) {
     return null;
   }
 }
+
+export function checkAscii(str: string) {
+  for (let i = 0; i < str.length; i++) {
```
Comment (Jarred-Sumner) [inline src/js/_codegen/helpers.ts]: `Buffer.isAscii` in `node:buffer`

File: src/js/builtins/ConsoleObject.ts [hunk 7]
```diff
@@ -880,5 +880,6 @@ export function createConsoleConstructor(console: typeof globalThis.console) {
   Console.prototype.error = Console.prototype.warn;
   Console.prototype.groupCollapsed = Console.prototype.group;
 
+  console.log("loaded console");
```
Comment (Jarred-Sumner) [inline src/js/builtins/ConsoleObject.ts]: ✂️

File: src/js/internal/primordials.js [hunk 8]
```diff
@@ -20,22 +21,16 @@ const createSafeIterator = (factory, next) => {
 };
 
 function getGetter(cls, getter) {
-  // TODO: __lookupGetter__ is deprecated, but Object.getOwnPropertyDescriptor doesn't work on built-ins like Typed Arrays.
-  return Function.prototype.call.bind(cls.prototype.__lookupGetter__(getter));
+  var g = cls.prototype.__lookupGetter__(getter);
```
Comment (Jarred-Sumner) [inline src/js/internal/primordials.js]: wat

File: test/js/node/stubs.test.js [hunk 9]
```diff
@@ -77,14 +77,17 @@ specifiers = [...weirdInternalSpecifiers, ...specifiers.flatMap(a => ["node:" +
 
 for (let specifier of specifiers) {
   test(`stubbed CJS import.meta.require ${specifier}`, async () => {
+    console.log(`importing1 ${specifier}`);
```
Comment (Jarred-Sumner) [inline test/js/node/stubs.test.js]: ✂️

File: test/js/node/stubs.test.js [hunk 10]
```diff
@@ -77,14 +77,17 @@ specifiers = [...weirdInternalSpecifiers, ...specifiers.flatMap(a => ["node:" +
...
     import.meta.require(specifier);
   });
 
   test(`stubbed CJS require ${specifier}`, async () => {
+    console.log(`importing2 ${specifier}`);
```
Comment (Jarred-Sumner) [inline test/js/node/stubs.test.js]: ✂️

File: test/js/node/stubs.test.js [hunk 11]
```diff
@@ -77,14 +77,17 @@ specifiers = [...weirdInternalSpecifiers, ...specifiers.flatMap(a => ["node:" +
...
     require(specifier);
   });
 
   test(`stubbed import ${specifier}`, async () => {
+    console.log(`importing3 ${specifier}`);
```
Comment (Jarred-Sumner) [inline test/js/node/stubs.test.js]: ✂️

File: test/js/node/util/node-inspect-tests/parallel/util-inspect.test.js [hunk 12]
```diff
@@ -1868,6 +1868,7 @@ test("no assertion failures 3", () => {
   assert.strictEqual(util.inspect("'"), '"\'"');
   assert.strictEqual(util.inspect("\"'"), "`\"'`");
   assert.strictEqual(util.inspect("\"'${a}"), "'\"\\'${a}'");
+  console.log(1);
```
Comment (Jarred-Sumner) [inline test/js/node/util/node-inspect-tests/parallel/util-inspect.test.js]: ✂️

File: test/js/node/util/node-inspect-tests/parallel/util-inspect.test.js [hunk 13]
```diff
@@ -3017,6 +3018,7 @@ test("no assertion failures 3", () => {
     assert.strictEqual(util.inspect(x, { maxStringLength: 4 }), "'aaaa'... 999996 more characters");
     assert.match(util.inspect(x, { maxStringLength: null }), /a'$/);
   }
+  console.log(8);
```
Comment (Jarred-Sumner) [inline test/js/node/util/node-inspect-tests/parallel/util-inspect.test.js]: ✂️

File: test/js/node/util/node-inspect-tests/parallel/util-inspect.test.js [hunk 14]
```diff
@@ -3103,6 +3105,7 @@ test("no assertion failures 3", () => {
     // Consistency check.
     assert(fullObjectGraph(global).has(Function.prototype));
   }
+  console.log(9);
```
Comment (Jarred-Sumner) [inline test/js/node/util/node-inspect-tests/parallel/util-inspect.test.js]: ✂️

File: test/transpiler/decorators.test.ts [hunk 15]
```diff
@@ -729,7 +729,7 @@ test("decorators with different property key types", () => {
       "string" = 30;
       @d1("string method")
       "string method"() {}
-      @d1(12e3)
+      @d1(12000)
```
Comment (Jarred-Sumner) [inline test/transpiler/decorators.test.ts:732]: why?

Comment (dylan-conway) [reply test/transpiler/decorators.test.ts:732]: I think that fixed a broken test a few days ago lol

File: src/bun.js/bindings/JSBufferList.cpp (hunk 1 above)
Comment (paperclover) [reply src/bun.js/bindings/JSBufferList.cpp:134]: no

```cpp
    if (total <= 0 || length() == 0) {
        RELEASE_AND_RETURN(throwScope, JSC::jsEmptyString(vm));
    }
```

File: src/deps/lol-html.zig (hunk 5 above)
Comment (paperclover) [reply src/deps/lol-html.zig:585]: Yes, it is an assertion failure to use createExternal on a zero-length string

File: src/js/internal/primordials.js (hunk 8 above)
Comment (paperclover) [reply src/js/internal/primordials.js]: The changes in this file are due to JSC not allowing the identifiers `call` and `apply` in a debug build. I will commit a better workaround

**Comment by Jarred-Sumner** [comment]
lets merge after tests run

**Pull request merged**